app/assets/javascripts/location_search.js

```diff
@@ -0,0 +1,73 @@
+window.ST = window.ST || {};
+
+(function(module) {
+
+  module.initializeLocationSearch = function() {
+    var input = document.getElementById('q');
+    var autocomplete = new window.google.maps.places.Autocomplete(input);
+    autocomplete.setTypes(['geocode']);
+
+    window.google.maps.event.addListener(autocomplete, 'place_changed', function(){
+      var place = autocomplete.getPlace();
+      if(place != null) {
+        if(place.geometry != null) {
+          document.getElementById('lc').value = place.geometry.location.toUrlValue();
+          document.getElementById('ls').value = window.google.maps.places.PlacesServiceStatus.OK;
+          document.getElementById('homepage-filters').submit();
+        } else {
+          // Let's pick first suggestion, if no geometry was returned by autocompletion
+          queryPredictions(place.name, handlePredictions);
+        }
+      }
+    });
+
+    // Ensure default events don't fire without correct info
+    var homepageForm = document.getElementById('homepage-filters');
+    homepageForm.addEventListener('submit', function(e) {
+      // If service status is unset and there are no coordinates, do not make search submit
+      if(document.getElementById('ls').value === "" && document.getElementById('lc').value === "") {
+        e.preventDefault();
+      }
+      // Submit will be triggered again after call to queryPredictions()
+      queryPredictions(input.value, handlePredictions);
+    });
+
+
+    var queryPredictions = function(inputString, callback) {
+      var autocompleteService = new window.google.maps.places.AutocompleteService();
+      autocompleteService.getQueryPredictions({ input: inputString }, callback);
+    };
+
+    var handlePredictions = function(predictions, status) {
+      var serviceStatus = window.google.maps.places.PlacesServiceStatus;
+
+      if(status === serviceStatus.OK) {
+        var map = new window.google.maps.Map(document.createElement('div'));
+        var placeService = new window.google.maps.places.PlacesService(map);
+
+        placeService.getDetails({
+          placeId: predictions[0].place_id // first prediction is default
+        }, function(place, status) {
+
+          if(status === serviceStatus.OK) {
+            document.getElementById('lc').value = place.geometry.location.toUrlValue();
+            document.getElementById('ls').value = serviceStatus.OK;
+            document.getElementById('homepage-filters').submit();
+          } else {
+            // Save received service status for logging
+            document.getElementById('ls').value = status;
+            document.getElementById('homepage-filters').submit();
+          }
+
+        });
+      } else {
+        // Save received service status for logging
+        document.getElementById('ls').value = status;
+        document.getElementById('homepage-filters').submit();
```

Comment:
All these code branches seem to end up submitting the form, so maybe you could take the `submit()` away from all the `if-else` branches and just do it at the end of `handlePredictions` function and that way reduce repetition?